### ISSUE TITLE
Errors when loading 500+ resources with `ext_resource` from zip 

### ISSUE THREAD

**Issue by zaevi** [godotengine/godot#42329]
**Godot version:**
3.2.3.stable.mono

**OS/device including version:**
Windows 10 

**Issue description:**
Currently I use exported zip for testing. When the project loads several resources from zip, the engine will throw errors like this:

```
.............
load 246.tres
load 247.tres
load 248.tres
load 249.tres
load 250.tres
load 251.tres
ERROR: Cannot open file 'dlc.zip'.
   At: core/io/file_access_zip.cpp:133
ERROR: _open: Condition "!zfile" is true. Returned: FAILED
   At: core/io/file_access_zip.cpp:269
ERROR: get_buffer: Condition "!zfile" is true. Returned: -1
   At: core/io/file_access_zip.cpp:333
ERROR: eof_reached: Condition "!zfile" is true. Returned: true
   At: core/io/file_access_zip.cpp:319
ERROR: open: res://resource/252_ext.tres:1 - Parse Error:
   At: scene/resources/resource_format_text.cpp:843
ERROR: Failed loading resource: res://resource/252_ext.tres. Make sure resources have been imported by opening the project in the editor at least once.
   At: core/io/resource_loader.cpp:278
ERROR: poll: res://resource/252.tres:3 - Parse Error: [ext_resource] referenced nonexistent resource at: res://resource/252_ext.tres
   At: scene/resources/resource_format_text.cpp:440
ERROR: Failed to load resource 'res://resource/252.tres'.
   At: core/io/resource_loader.cpp:208
ERROR: Failed loading resource: res://resource/252.tres. Make sure resources have been imported by opening the project in the editor at least once.
   At: core/io/resource_loader.cpp:278
ERROR: Error loading resource: 'resource/252.tres'.
   At: core/bind/core_bind.cpp:76
```

**This will happen when:
1.load resources from zip
2.resource contains `ext_resource` to load
3.a lot of resources has been loaded** 

Sorry for my bad English, and it's not easy to describe this strange issue. In my actual project, I export many AtlasTextures with its original textures.

Other information:
1.it's ok for pck
2.it's ok to load resources without any `ext_resource`
3.it seems failed after 500 resources (including ext) has been loaded. If the tres had two externals, it would throw errors after the 167th.

**Steps to reproduce:**
You should reproduce it using demo project below.
1. generate 300 external resources `(0_ext.tres - 299_ext.tres)` and 300 resources `(0.tres - 299.tres)`.
2. export them as zip.
3. load the zip, and try loading from `0.tres` to `299.tres`.

**Minimal reproduction project:**
[zip_bug.zip](https://github.com/godotengine/godot/files/5280979/zip_bug.zip)

**Comment by akien-mga** [comment]
I can't reproduce it on Linux with 3.2.3-stable (the `demo` project in the attached zip works fine for me, it loads all 300 resources + 300 external resources), so I assume it might be a Windows-specific issue.

**Comment by akien-mga** [comment]
Can't get `winedbg --gdb` to load debug symbols for my MinGW binary, so here's a quick print-based debugging attempt:
```diff


        diff --git a/core/io/file_access_zip.cpp b/core/io/file_access_zip.cpp
        index 6120146e1b..6c4f82bf06 100644
        --- a/core/io/file_access_zip.cpp
        +++ b/core/io/file_access_zip.cpp
        @@ -129,8 +129,9 @@ unzFile ZipArchive::get_file_handle(String p_file) const {
                ERR_FAIL_COND_V_MSG(!file_exists(p_file), NULL, "File '" + p_file + " doesn't exist.");
                File file = files[p_file];
         
        -       FileAccess *f = FileAccess::open(packages[file.package].filename, FileAccess::READ);
        -       ERR_FAIL_COND_V_MSG(!f, NULL, "Cannot open file '" + packages[file.package].filename + "'.");
        +       Error err;
        +       FileAccess *f = FileAccess::open(packages[file.package].filename, FileAccess::READ, &err);
        +       ERR_FAIL_COND_V_MSG(!f, NULL, "Cannot open file '" + packages[file.package].filename + "'. Error code: " + itos(err) + ".");
         
                zlib_filefunc_def io;
                zeromem(&io, sizeof(io));
        diff --git a/drivers/windows/file_access_windows.cpp b/drivers/windows/file_access_windows.cpp
        index 1b8760fdc7..e236786a81 100644
        --- a/drivers/windows/file_access_windows.cpp
        +++ b/drivers/windows/file_access_windows.cpp
        @@ -116,6 +116,7 @@ Error FileAccessWindows::_open(const String &p_path, int p_mode_flags) {
                }
         
                errno_t errcode = _wfopen_s(&f, path.c_str(), mode_string);
        +       print_line("File: " + path + ". Errcode from _wfopen_s: " + itos((int)errcode));
         
                if (f == NULL) {
                        switch (errcode) {
         
```

Prints:
```
...
load 249.tres
File: D:/tmp/godot/bug/zip_bug/demo/resource/250.tres.remap. Errcode from _wfopen_s: 2
File: dlc.zip. Errcode from _wfopen_s: 0
File: dlc.zip. Errcode from _wfopen_s: 0
File: D:/tmp/godot/bug/zip_bug/demo/resource/250_ext.tres.remap. Errcode from _wfopen_s: 2
File: dlc.zip. Errcode from _wfopen_s: 0
File: dlc.zip. Errcode from _wfopen_s: 0
load 250.tres
File: D:/tmp/godot/bug/zip_bug/demo/resource/251.tres.remap. Errcode from _wfopen_s: 2
File: dlc.zip. Errcode from _wfopen_s: 0
File: dlc.zip. Errcode from _wfopen_s: 0
File: D:/tmp/godot/bug/zip_bug/demo/resource/251_ext.tres.remap. Errcode from _wfopen_s: 2
File: dlc.zip. Errcode from _wfopen_s: 0
File: dlc.zip. Errcode from _wfopen_s: 0
load 251.tres
File: D:/tmp/godot/bug/zip_bug/demo/resource/252.tres.remap. Errcode from _wfopen_s: 2
File: dlc.zip. Errcode from _wfopen_s: 0
File: dlc.zip. Errcode from _wfopen_s: 0
File: D:/tmp/godot/bug/zip_bug/demo/resource/252_ext.tres.remap. Errcode from _wfopen_s: 2
File: dlc.zip. Errcode from _wfopen_s: 2
ERROR: Cannot open file 'dlc.zip'. Error code: 7.
   At: core/io/file_access_zip.cpp:134
```

So at some point it just decides that `dlc.zip` is `ENOENT` (not found).

Maybe we don't close file handles properly and we run out of them?

**Comment by akien-mga** [comment]
Yep Windows has a max of 512 file descriptors: https://docs.microsoft.com/en-us/cpp/c-runtime-library/file-handling?view=vs-2019

This probably matches the ~500 resources needed to start failing + the other files from the project.

**Comment by zaevi** [comment]
Confirmed.
![handle](https://user-images.githubusercontent.com/12966814/94251702-7ba5fc00-ff55-11ea-8ae2-4fa1760c5a38.png)

But I found when loading empty resources (more than 512) there are still handles remain and no error throwed. Seems there is a cache for zip file descriptors (at least it will close previous handles)

Then I generated 600 resources + 600 external resources for testing:
```
# load 600 ext resources first (empty)
for i in range(600):
    if ResourceLoader.load("resource/%d_ext.tres" % i):
        prints("load", "%d_ext.tres" % i)
    else:
        break
# ok and 500+ handles remain

# load a res with ext_resource
ResourceLoader.load("resource/1.tres") # throw errors
```

So errors causes when file handles are full and you are loading a res with external res.

**Comment by zaevi** [comment]
Alright, this is because `ZipArchive::get_file_handle` opens zip **TWICE** (line 120 & 139)
https://github.com/godotengine/godot/blob/96a51b108c953d8c8e112340b03fc1026d7243ed/core/io/file_access_zip.cpp#L116-L148
The second opening is unnecessary and not closed (line 48)
https://github.com/godotengine/godot/blob/96a51b108c953d8c8e112340b03fc1026d7243ed/core/io/file_access_zip.cpp#L42-L51

This is not windows-related and you could find handles increasing in other platforms.

After removing line 48 the tests run well and there are no extra handles.

**Comment by Host32** [comment]
I have a siminar error here.

I created a launcher for mobile that dowloads the game updated patches as .zip and the error appears intermittently on some android smartphones.

```
**ERROR**: Cannot open file 'user://downloaded_packs/pack.Game.Battle.BattleEntity.BattleComponent.AttributesComponent.zip'.
   At: core/io/file_access_zip.cpp:133:get_file_handle() - Condition "!f" is true. Returned: __null
**ERROR**: Condition "!zfile" is true. Returned: FAILED
   At: core/io/file_access_zip.cpp:272:_open() - Condition "!zfile" is true. Returned: FAILED
**ERROR**: Condition "!zfile" is true. Returned: -1
   At: core/io/file_access_zip.cpp:336:get_buffer() - Condition "!zfile" is true. Returned: -1
**ERROR**: Condition "!zfile" is true. Returned: true
   At: core/io/file_access_zip.cpp:322:eof_reached() - Condition "!zfile" is true. Returned: true
**ERROR**: Condition "err" is true. Returned: err
   At: modules/gdscript/gdscript.cpp:829:load_source_code() - Condition "err" is true. Returned: err
**ERROR**: Cannot load source code from file 'res://Game/Battle/BattleEntity/BattleComponent/AttributesComponent/AttributesComponent.gd'.
   At: modules/gdscript/gdscript.cpp:2315:load() - Condition "err != OK" is true. Returned: RES()
**ERROR**: Failed loading resource: res://Game/Battle/BattleEntity/BattleComponent/AttributesComponent/AttributesComponent.gd. Make sure resources have been imported by opening the project in the editor at least once.
   At: core/io/resource_loader.cpp:279:_load() - Condition "found" is true. Returned: RES()
**ERROR**: res://Game/Battle/BattleEntity/BattleComponent/AttributesComponent/AttributesComponent.tscn:3 - Parse Error: [ext_resource] referenced nonexistent resource at: res://Game/Battle/BattleEntity/BattleComponent/AttributesComponent/AttributesComponent.gd
   At: scene/resources/resource_format_text.cpp:440:poll() - res://Game/Battle/BattleEntity/BattleComponent/AttributesComponent/AttributesComponent.tscn:3 - Parse Error: [ext_resource] referenced nonexistent resource at: res://Game/Battle/BattleEntity/BattleComponent/AttributesComponent/AttributesComponent.gd
**ERROR**: Failed to load resource 'res://Game/Battle/BattleEntity/BattleComponent/AttributesComponent/AttributesComponent.tscn'.
   At: core/io/resource_loader.cpp:208:load() - Condition "err != OK" is true. Returned: RES()
**ERROR**: Failed loading resource: res://Game/Battle/BattleEntity/BattleComponent/AttributesComponent/AttributesComponent.tscn. Make sure resources have been imported by opening the project in the editor at least once.
   At: core/io/resource_loader.cpp:279:_load() - Condition "found" is true. Returned: RES()
**ERROR**: res://Game/Battle/BattleEntity/BattleEntityChampion/BattleEntityChampion.tscn:15 - Parse Error: [ext_resource] referenced nonexistent resource at: res://Game/Battle/BattleEntity/BattleComponent/AttributesComponent/AttributesComponent.tscn
   At: scene/resources/resource_format_text.cpp:440:poll() - res://Game/Battle/BattleEntity/BattleEntityChampion/BattleEntityChampion.tscn:15 - Parse Error: [ext_resource] referenced nonexistent resource at: res://Game/Battle/BattleEntity/BattleComponent/AttributesComponent/AttributesComponent.tscn
**ERROR**: Failed to load resource 'res://Game/Battle/BattleEntity/BattleEntityChampion/BattleEntityChampion.tscn'.
   At: core/io/resource_loader.cpp:208:load() - Condition "err != OK" is true. Returned: RES()
**ERROR**: Failed loading resource: res://Game/Battle/BattleEntity/BattleEntityChampion/BattleEntityChampion.tscn. Make sure resources have been imported by opening the project in the editor at least once.
   At: core/io/resource_loader.cpp:279:_load() - Condition "found" is true. Returned: RES()
**SCRIPT ERROR**: Parse Error: Can't preload resource at path: res://Game/Battle/BattleEntity/BattleEntityChampion/BattleEntityChampion.tscn
   At: res://Game/Battle/Battle.gdc:9:GDScript::load_byte_code() - Parse Error: Can't preload resource at path: res://Game/Battle/BattleEntity/BattleEntityChampion/BattleEntityChampion.tscn
**ERROR**: Method failed. Returning: ERR_PARSE_ERROR
   At: modules/gdscript/gdscript.cpp:801:load_byte_code() - Method failed. Returning: ERR_PARSE_ERROR
**ERROR**: Cannot load byte code from file 'res://Game/Battle/Battle.gdc'.
   At: modules/gdscript/gdscript.cpp:2311:load() - Condition "err != OK" is true. Returned: RES()
**ERROR**: Failed loading resource: res://Game/Battle/Battle.gdc. Make sure resources have been imported by opening the project in the editor at least once.
   At: core/io/resource_loader.cpp:279:_load() - Condition "found" is true. Returned: RES()
**ERROR**: res://Game/Battle/Battle.tscn:7 - Parse Error: [ext_resource] referenced nonexistent resource at: res://Game/Battle/Battle.gd
   At: scene/resources/resource_format_text.cpp:440:poll() - res://Game/Battle/Battle.tscn:7 - Parse Error: [ext_resource] referenced nonexistent resource at: res://Game/Battle/Battle.gd
**ERROR**: Failed to load resource 'res://Game/Battle/Battle.tscn'.
   At: core/io/resource_loader.cpp:208:load() - Condition "err != OK" is true. Returned: RES()
**ERROR**: Failed loading resource: res://Game/Battle/Battle.tscn. Make sure resources have been imported by opening the project in the editor at least once.
   At: core/io/resource_loader.cpp:279:_load() - Condition "found" is true. Returned: RES()
**SCRIPT ERROR**: Parse Error: Can't preload resource at path: res://Game/Battle/Battle.tscn
   At: res://Game/WorldMap/Navigable/World/Empire/BattlePoint/BattlePointUI/BattlePointUI.gdc:5:GDScript::load_byte_code() - Parse Error: Can't preload resource at path: res://Game/Battle/Battle.tscn
**ERROR**: Method failed. Returning: ERR_PARSE_ERROR
   At: modules/gdscript/gdscript.cpp:801:load_byte_code() - Method failed. Returning: ERR_PARSE_ERROR
**ERROR**: Cannot load byte code from file 'res://Game/WorldMap/Navigable/World/Empire/BattlePoint/BattlePointUI/BattlePointUI.gdc'.
   At: modules/gdscript/gdscript.cpp:2311:load() - Condition "err != OK" is true. Returned: RES()
**ERROR**: Failed loading resource: res://Game/WorldMap/Navigable/World/Empire/BattlePoint/BattlePointUI/BattlePointUI.gdc. Make sure resources have been imported by opening the project in the editor at least once.
   At: core/io/resource_loader.cpp:279:_load() - Condition "found" is true. Returned: RES()
**ERROR**: res://Game/WorldMap/Navigable/World/Empire/BattlePoint/BattlePointUI/BattlePointUI.tscn:3 - Parse Error: [ext_resource] referenced nonexistent resource at: res://Game/WorldMap/Navigable/World/Empire/BattlePoint/BattlePointUI/BattlePointUI.gd
   At: scene/resources/resource_format_text.cpp:440:poll() - res://Game/WorldMap/Navigable/World/Empire/BattlePoint/BattlePointUI/BattlePointUI.tscn:3 - Parse Error: [ext_resource] referenced nonexistent resource at: res://Game/WorldMap/Navigable/World/Empire/BattlePoint/BattlePointUI/BattlePointUI.gd
```

**Comment by Host32** [comment]
I have a similar problem with Android 8 and previous.

I created a launcher for my mobile game that downloads zipped packs and use `ProjectSettings.load_resource_pack()` to add it to `res://` dir.

So, when the player enters a place in the game, i use `ResourceLoader.load_interactive()` to load the current scene that was previous loaded in a resource pack.

On Android 9 and later, it works correctly, but on Android 8 and earlier, the `poll` method of interactive loader returns `ERR_FILE_CORRUPT`, and in console i get the following error.

I also checked the md5 of the .zip file and its not corrupted.

Is this error related with this issue?

```
**ERROR**: Cannot open file 'user://downloaded_packs/pack.Game.Battle.BattleEntity.BattleComponent.FloatingBoxComponent.AnimatedText.zip'.
   At: core/io/file_access_zip.cpp:133:get_file_handle() - Condition "!f" is true. Returned: __null
**ERROR**: Condition "!zfile" is true. Returned: FAILED
   At: core/io/file_access_zip.cpp:272:_open() - Condition "!zfile" is true. Returned: FAILED
**ERROR**: Condition "!zfile" is true. Returned: -1
   At: core/io/file_access_zip.cpp:336:get_buffer() - Condition "!zfile" is true. Returned: -1
**ERROR**: Condition "!zfile" is true. Returned: true
   At: core/io/file_access_zip.cpp:322:eof_reached() - Condition "!zfile" is true. Returned: true
**ERROR**: Condition "err" is true. Returned: err
   At: modules/gdscript/gdscript.cpp:829:load_source_code() - Condition "err" is true. Returned: err
**ERROR**: Cannot load source code from file 'res://Game/Battle/BattleEntity/BattleComponent/FloatingBoxComponent/AnimatedText/AnimatedText.gd'.
   At: modules/gdscript/gdscript.cpp:2315:load() - Condition "err != OK" is true. Returned: RES()
**ERROR**: Failed loading resource: res://Game/Battle/BattleEntity/BattleComponent/FloatingBoxComponent/AnimatedText/AnimatedText.gd. Make sure resources have been imported by opening the project in the editor at least once.
   At: core/io/resource_loader.cpp:279:_load() - Condition "found" is true. Returned: RES()
**ERROR**: res://Game/Battle/BattleEntity/BattleComponent/FloatingBoxComponent/AnimatedText/AnimatedText.tscn:3 - Parse Error: [ext_resource] referenced nonexistent resource at: res://Game/Battle/BattleEntity/BattleComponent/FloatingBoxComponent/AnimatedText/AnimatedText.gd
   At: scene/resources/resource_format_text.cpp:440:poll() - res://Game/Battle/BattleEntity/BattleComponent/FloatingBoxComponent/AnimatedText/AnimatedText.tscn:3 - Parse Error: [ext_resource] referenced nonexistent resource at: res://Game/Battle/BattleEntity/BattleComponent/FloatingBoxComponent/AnimatedText/AnimatedText.gd
**ERROR**: Failed to load resource 'res://Game/Battle/BattleEntity/BattleComponent/FloatingBoxComponent/AnimatedText/AnimatedText.tscn'.
   At: core/io/resource_loader.cpp:208:load() - Condition "err != OK" is true. Returned: RES()
**ERROR**: Failed loading resource: res://Game/Battle/BattleEntity/BattleComponent/FloatingBoxComponent/AnimatedText/AnimatedText.tscn. Make sure resources have been imported by opening the project in the editor at least once.
   At: core/io/resource_loader.cpp:279:_load() - Condition "found" is true. Returned: RES()
**SCRIPT ERROR**: Parse Error: Can't preload resource at path: res://Game/Battle/BattleEntity/BattleComponent/FloatingBoxComponent/AnimatedText/AnimatedText.tscn
   At: res://Game/Battle/BattleEntity/BattleComponent/FloatingBoxComponent/FloatingBoxComponent.gdc:8:GDScript::load_byte_code() - Parse Error: Can't preload resource at path: 
```

**Comment by avnerh1** [comment]
What is the workaround, guys? Switching from .zip to .pck and waiting for Godot 4.0?

I have the same issue with iOS.

**Comment by zaevi** [comment]
> What is the workaround, guys? Switching from .zip to .pck and waiting for Godot 4.0?
> 
> I have the same issue with iOS.

This have been fixed in latest stable (3.3.2) and I'm not sure if your issue is related to this. You can open issue to describe in detail.

**Comment by avnerh1** [comment]
I'm still using 3.3. Will try with 3.3.2
